File: backend/app/services/profile_service.py
from app.models.profile import ProfileModel
from app.utils.validators import validate_disability_type, validate_severity
from app.utils.helpers import serialize_doc


class ProfileService:

    @staticmethod
    def create_profile(user_id, data):
        existing = ProfileModel.get_by_user_id(user_id)
        if existing:
            return {'error': 'Profile already exists. Use PUT to update.'}, 409

        disabilities = data.get('disabilities', [])
        for d in disabilities:
            if not validate_disability_type(d.get('type', '')):
                return {'error': f"Invalid disability type: {d.get('type')}"}, 400
            if not validate_severity(d.get('severity', '')):
                return {'error': f"Invalid severity: {d.get('severity')}"}, 400

        profile_id = ProfileModel.create_profile(user_id, data)
        profile = ProfileModel.get_by_user_id(user_id)

        return {
            'message': 'Profile created successfully',
            'profile': serialize_doc(profile),
        }, 201

    @staticmethod
    def get_profile(user_id):
        profile = ProfileModel.get_by_user_id(user_id)
        if not profile:
            return {'error': 'Profile not found'}, 404
        return {'profile': serialize_doc(profile)}, 200

    @staticmethod
    def update_profile(user_id, data):
        existing = ProfileModel.get_by_user_id(user_id)
        if not existing:
            return {'error': 'Profile not found'}, 404

        disabilities = data.get('disabilities', [])
        for d in disabilities:
            if not validate_disability_type(d.get('type', '')):
                return {'error': f"Invalid disability type: {d.get('type')}"}, 400
            if not validate_severity(d.get('severity', '')):
                return {'error': f"Invalid severity: {d.get('severity')}"}, 400

        ProfileModel.update_profile(user_id, data)
        profile = ProfileModel.get_by_user_id(user_id)

        return {
            'message': 'Profile updated successfully',
            'profile': serialize_doc(profile),
        }, 200

    @staticmethod
    def update_preferences(user_id, preferences):
        existing = ProfileModel.get_by_user_id(user_id)
        if not existing:
            return {'error': 'Profile not found'}, 404

        ProfileModel.update_preferences(user_id, preferences)
        profile = ProfileModel.get_by_user_id(user_id)

        return {
            'message': 'Preferences updated successfully',
            'profile': serialize_doc(profile),
        }, 200

File: frontend/src/services/profileService.js
import { getStore, setStore, generateId, delay, getCurrentUserId, KEYS } from './mockData'

export const createProfile = async (data) => {
  await delay(300)
  const userId = getCurrentUserId()
  if (!userId) throw { response: { status: 401, data: { error: 'Not authenticated' } } }

  const profiles = getStore(KEYS.PROFILES) || []
  const existing = profiles.find((p) => p.user_id === userId)
  if (existing) throw { response: { status: 409, data: { error: 'Profile already exists. Use PUT to update.' } } }

  const profile = {
    _id: generateId(),
    user_id: userId,
    disabilities: data.disabilities || [],
    preferences: data.preferences || {},
    location: data.location || {},
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString(),
  }

  profiles.push(profile)
  setStore(KEYS.PROFILES, profiles)

  return { data: { profile, message: 'Profile created successfully' } }
}

export const getProfile = async () => {
  await delay(200)
  const userId = getCurrentUserId()
  if (!userId) throw { response: { status: 401, data: { error: 'Not authenticated' } } }

  const profiles = getStore(KEYS.PROFILES) || []
  const profile = profiles.find((p) => p.user_id === userId)
  if (!profile) throw { response: { status: 404, data: { error: 'Profile not found' } } }

  return { data: { profile } }
}

export const updateProfile = async (data) => {
  await delay(300)
  const userId = getCurrentUserId()
  if (!userId) throw { response: { status: 401, data: { error: 'Not authenticated' } } }

  const profiles = getStore(KEYS.PROFILES) || []
  const idx = profiles.findIndex((p) => p.user_id === userId)
  if (idx === -1) throw { response: { status: 404, data: { error: 'Profile not found' } } }

  profiles[idx] = {
    ...profiles[idx],
    disabilities: data.disabilities ?? profiles[idx].disabilities,
    preferences: data.preferences ?? profiles[idx].preferences,
    location: data.location ?? profiles[idx].location,
    updated_at: new Date().toISOString(),
  }
  setStore(KEYS.PROFILES, profiles)

  return { data: { profile: profiles[idx], message: 'Profile updated successfully' } }
}

export const updatePreferences = async (preferences) => {
  await delay(200)
  const userId = getCurrentUserId()
  if (!userId) throw { response: { status: 401, data: { error: 'Not authenticated' } } }

  const profiles = getStore(KEYS.PROFILES) || []
  const idx = profiles.findIndex((p) => p.user_id === userId)
  if (idx === -1) throw { response: { status: 404, data: { error: 'Profile not found' } } }

  profiles[idx].preferences = { ...profiles[idx].preferences, ...preferences }
  profiles[idx].updated_at = new Date().toISOString()
  setStore(KEYS.PROFILES, profiles)

  return { data: { profile: profiles[idx], message: 'Preferences updated successfully' } }
}
